File: java-sdk/src/main/java/org/arguslog/sdk/Dsn.java
package org.arguslog.sdk;

import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public record Dsn(
    String publicKey, String host, String scheme, String projectId, String ingestUrl) {

  // User-facing DSN format (matches what the api emits in the project-create response and shows
  // in the web UI's "Copy DSN" modal): arguslog://<publicKey>@<host>/api/<projectId>
  // Kept in lockstep with packages/sdk-browser/src/dsn.ts.
  private static final Pattern DSN_RE =
      Pattern.compile("^arguslog://([^@]+)@([^/]+)/api/([^/?#]+)$");

  public static Dsn parse(String raw) {
    Objects.requireNonNull(raw, "dsn");
    Matcher m = DSN_RE.matcher(raw);
    if (!m.matches()) {
      throw new IllegalArgumentException("Invalid DSN: " + raw);
    }
    String publicKey = m.group(1);
    String host = m.group(2);
    String projectId = m.group(3);
    if (publicKey.isEmpty() || host.isEmpty() || projectId.isEmpty()) {
      throw new IllegalArgumentException("Invalid DSN: " + raw);
    }
    String scheme = isDevHost(host) ? "http" : "https";
    String ingestUrl = scheme + "://" + host + "/api/" + projectId + "/events";
    return new Dsn(publicKey, host, scheme, projectId, ingestUrl);
  }

  private static boolean isDevHost(String host) {
    String bare =
        host.startsWith("[") ? host.substring(0, host.indexOf(']') + 1) : host.split(":", 2)[0];
    if (bare.equals("localhost")
        || bare.startsWith("127.")
        || bare.equals("0.0.0.0")
        || bare.equals("[::1]")
        || bare.equals("::1")) {
      return true;
    }
    // RFC1918 private ranges: 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16. A device on the same
    // LAN pointing at the dev box's LAN IP is still a dev-mode transport; ingest only listens
    // on plain HTTP locally, so an HTTPS upgrade would fail the TLS handshake silently.
    String[] parts = bare.split("\\.");
    if (parts.length != 4) return false;
    int a, b;
    try {
      a = Integer.parseInt(parts[0]);
      b = Integer.parseInt(parts[1]);
      Integer.parseInt(parts[2]);
      Integer.parseInt(parts[3]);
    } catch (NumberFormatException ex) {
      return false;
    }
    if (a == 10) return true;
    if (a == 192 && b == 168) return true;
    if (a == 172 && b >= 16 && b <= 31) return true;
    return false;
  }
}

File: python-sdk/src/arguslog/_dsn.py
"""DSN parser kept in lockstep with packages/sdk-browser/src/dsn.ts and
java-sdk/src/main/java/org/arguslog/sdk/Dsn.java.

User-facing DSN format: arguslog://<publicKey>@<host>/api/<projectId>
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_DSN_RE = re.compile(r"^arguslog://([^@]+)@([^/]+)/api/([^/?#]+)$")
_LOOPBACK_HOSTS = {"localhost", "0.0.0.0", "[::1]", "::1"}


class InvalidDsnError(ValueError):
    """Raised when a DSN string fails to match the expected shape."""


@dataclass(frozen=True)
class ParsedDsn:
    public_key: str
    host: str
    scheme: str
    project_id: str
    ingest_url: str


def parse_dsn(raw: str) -> ParsedDsn:
    if raw is None:
        raise InvalidDsnError("dsn is required")
    match = _DSN_RE.match(raw)
    if not match:
        raise InvalidDsnError(f"Invalid DSN: {raw}")
    public_key, host, project_id = match.group(1), match.group(2), match.group(3)
    if not public_key or not host or not project_id:
        raise InvalidDsnError(f"Invalid DSN: {raw}")
    scheme = "http" if _is_dev_host(host) else "https"
    ingest_url = f"{scheme}://{host}/api/{project_id}/events"
    return ParsedDsn(public_key, host, scheme, project_id, ingest_url)


def _is_dev_host(host: str) -> bool:
    bare = host[: host.index("]") + 1] if host.startswith("[") else host.split(":", 1)[0]
    if bare in _LOOPBACK_HOSTS or bare.startswith("127."):
        return True
    # RFC1918 private ranges: 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16. A device on the same
    # LAN pointing at the dev box is still a dev-mode transport; ingest only listens on plain
    # HTTP locally, so HTTPS upgrade would fail the TLS handshake silently.
    parts = bare.split(".")
    if len(parts) != 4:
        return False
    try:
        a, b, c, d = (int(p) for p in parts)
    except ValueError:
        return False
    if not all(0 <= n <= 255 for n in (a, b, c, d)):
        return False
    if a == 10:
        return True
    if a == 192 and b == 168:
        return True
    return a == 172 and 16 <= b <= 31
